File: midolman/src/main/java/org/midonet/cluster/data/neutron/NetworkApi.java
/*
 * Copyright (c) 2014 Midokura SARL, All Rights Reserved.
 */
package org.midonet.cluster.data.neutron;

import org.midonet.cluster.data.Rule;
import org.midonet.midolman.serialization.SerializationException;
import org.midonet.midolman.state.StateAccessException;
import org.midonet.midolman.state.zkManagers.BridgeZkManager;

import javax.annotation.Nonnull;
import java.util.List;
import java.util.UUID;


public interface NetworkApi {

    /**
     * Create a new network data in the data store. StatePathExistsException
     * thrown if a network with the same ID already exists.
     *
     * @param network Network object to create
     * @return Created Network object
     */
    Network createNetwork(@Nonnull Network network)
            throws StateAccessException, SerializationException;

    /**
     * Create multiple networks atomically.
     *
     * @param networks Network objects to create
     * @return Created Network objects
     */
    List<Network> createNetworkBulk(@Nonnull List<Network> networks)
            throws StateAccessException, SerializationException;

    /**
     * Delete a network. Nothing happens if the resource does not exist.
     *
     * @param id ID of the Network object to delete
     */
    void deleteNetwork(@Nonnull UUID id)
            throws StateAccessException, SerializationException;

    /**
     * Retrieve a network. Returns null if the resource does not exist.
     *
     * @param id ID of the Network object to get
     * @return Network object
     */
    Network getNetwork(@Nonnull UUID id)
            throws StateAccessException, SerializationException;

    /**
     * Get all the networks.
     *
     * @return List of Network objects.
     */
    List<Network> getNetworks()
            throws StateAccessException, SerializationException;

    /**
     * Update a network. NoStatePathException is thrown if the resource does
     * not exist.
     *
     * @param id ID of the Network object to update
     * @return Updated Network object
     */
    Network updateNetwork(@Nonnull UUID id, @Nonnull Network network)
            throws StateAccessException, SerializationException,
            BridgeZkManager.VxLanPortIdUpdateException;

    /**
     * Create a new subnet data in the data store.  StatePathExistsException
     * thrown if the subnet with the same ID already exists.
     *
     * @param subnet Network object to create
     * @return Created Subnet object
     */
    Subnet createSubnet(@Nonnull Subnet subnet)
            throws StateAccessException, SerializationException;

    /**
     * Create multiple subnets atomically.
     *
     * @param subnets Subnet objects to create
     * @return Created Subnet objects
     */
    List<Subnet> createSubnetBulk(@Nonnull List<Subnet> subnets)
            throws StateAccessException, SerializationException;


    /**
     * Delete a subnet.  Nothing happens if the resource does not exist.
     *
     * @param id ID of the Subnet object to delete
     */
    void deleteSubnet(@Nonnull UUID id)
            throws StateAccessException, SerializationException;

    /**
     * Retrieve a subnet.  Returns null if the resource does not exist.
     *
     * @param id ID of the Subnet object to get
     * @return Subnet object
     */
    Subnet getSubnet(@Nonnull UUID id)
            throws StateAccessException, SerializationException;

    /**
     * Get all the subnets.
     *
     * @return List of Subnet objects.
     */
    List<Subnet> getSubnets()
            throws StateAccessException, SerializationException;

    /**
     * Update a subnet.  NoStatePathException is thrown if the resource does
     * not exist.
     *
     * @param id ID of the Subnet object to update
     * @return Updated Subnet object
     */
    Subnet updateSubnet(@Nonnull UUID id, @Nonnull Subnet subnet)
            throws StateAccessException, SerializationException;

    /**
     * Create a new port data in the data store. StatePathExistsException
     * thrown if the port with the same ID already exists.
     *
     * @param port port object to create
     * @return Created Port object
     */
    Port createPort(@Nonnull Port port)
            throws StateAccessException, SerializationException;

    /**
     * Create multiple ports atomically.
     *
     * @param ports Port objects to create
     * @return Created Port objects
     */
    List<Port>  createPortBulk(@Nonnull List<Port> ports)
            throws StateAccessException, SerializationException,
            Rule.RuleIndexOutOfBoundsException;
    /**
     * Delete a port. Nothing happens if the resource does not exist.
     *
     * @param id ID of the Port object to delete
     */
    void deletePort(@Nonnull UUID id)
            throws StateAccessException, SerializationException;

    /**
     * Retrieve a port. Returns null if the resource does not exist.
     *
     * @param id ID of the Port object to delete
     * @return Port object
     */
    Port getPort(@Nonnull UUID id)
            throws StateAccessException, SerializationException;

    /**
     * Get all the ports.
     *
     * @return List of Port objects.
     */
    List<Port> getPorts()
            throws StateAccessException, SerializationException;

    /**
     * Update a port. NoStatePathException is thrown if the resource does
     * not exist.
     *
     * @param id ID of the Port object to update
     * @return Updated Port object
     */
    Port updatePort(@Nonnull UUID id, @Nonnull Port port)
            throws StateAccessException, SerializationException,
            Rule.RuleIndexOutOfBoundsException;
}

File: midonet-cluster/src/main/scala/org/midonet/cluster/services/rest_api/neutron/plugin/NetworkApi.scala
/*
 * Copyright 2015 Midokura SARL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.midonet.cluster.services.rest_api.neutron.plugin

import java.util
import java.util.UUID
import javax.annotation.Nonnull

import org.midonet.cluster.rest_api.neutron.models.{Network, Port, Subnet}
import org.midonet.cluster.rest_api.{ConflictHttpException, NotFoundHttpException}

trait NetworkApi {

    /**
     * Create a new network data in the data store.
     *
     * @param network Network object to create
     * @return Created Network object
     */
    @throws(classOf[ConflictHttpException])
    @throws(classOf[NotFoundHttpException])
    def createNetwork(@Nonnull network: Network): Network

    /**
     * Create multiple networks atomically.
     *
     * @param networks Network objects to create
     * @return Created Network objects
     */
    @throws(classOf[ConflictHttpException])
    @throws(classOf[NotFoundHttpException])
    def createNetworkBulk(@Nonnull networks: util.List[Network])
    : util.List[Network]

    /**
     * Delete a network. Nothing happens if the resource does not exist.
     *
     * @param id ID of the Network object to delete
     */
    @throws(classOf[ConflictHttpException])
    @throws(classOf[NotFoundHttpException])
    def deleteNetwork(@Nonnull id: UUID)

    /**
     * Retrieve a network. Returns null if the resource does not exist.
     *
     * @param id ID of the Network object to get
     * @return Network object
     */
    @throws(classOf[ConflictHttpException])
    @throws(classOf[NotFoundHttpException])
    def getNetwork(@Nonnull id: UUID): Network

    /**
     * Get all the networks.
     *
     * @return util.List of Network objects.
     */
    @throws(classOf[ConflictHttpException])
    @throws(classOf[NotFoundHttpException])
    def getNetworks: util.List[Network]

    /**
     * Update a network.
     *
     * @param id ID of the Network object to update
     * @return Updated Network object
     */
    @throws(classOf[ConflictHttpException])
    @throws(classOf[NotFoundHttpException])
    def updateNetwork(@Nonnull id: UUID, @Nonnull network: Network): Network

    /**
     * Create a new subnet data in the data store.
     *
     * @param subnet Network object to create
     * @return Created Subnet object
     */
    @throws(classOf[ConflictHttpException])
    @throws(classOf[NotFoundHttpException])
    def createSubnet(@Nonnull subnet: Subnet): Subnet

    /**
     * Create multiple subnets atomically.
     *
     * @param subnets Subnet objects to create
     * @return Created Subnet objects
     */
    @throws(classOf[ConflictHttpException])
    @throws(classOf[NotFoundHttpException])
    def createSubnetBulk(@Nonnull subnets: util.List[Subnet])
    : util.List[Subnet]

    /**
     * Delete a subnet.  Nothing happens if the resource does not exist.
     *
     * @param id ID of the Subnet object to delete
     */
    @throws(classOf[ConflictHttpException])
    @throws(classOf[NotFoundHttpException])
    def deleteSubnet(@Nonnull id: UUID)

    /**
     * Retrieve a subnet.  Returns null if the resource does not exist.
     *
     * @param id ID of the Subnet object to get
     * @return Subnet object
     */
    @throws(classOf[ConflictHttpException])
    @throws(classOf[NotFoundHttpException])
    def getSubnet(@Nonnull id: UUID): Subnet

    /**
     * Get all the subnets.
     *
     * @return util.List of Subnet objects.
     */
    @throws(classOf[ConflictHttpException])
    @throws(classOf[NotFoundHttpException])
    def getSubnets: util.List[Subnet]

    /**
     * Update a subnet.
     *
     * @param id ID of the Subnet object to update
     * @return Updated Subnet object
     */
    @throws(classOf[ConflictHttpException])
    @throws(classOf[NotFoundHttpException])
    def updateSubnet(@Nonnull id: UUID, @Nonnull subnet: Subnet): Subnet

    /**
     * Create a new port data in the data store.
     *
     * @param port port object to create
     * @return Created Port object
     */
    @throws(classOf[ConflictHttpException])
    @throws(classOf[NotFoundHttpException])
    def createPort(@Nonnull port: Port): Port

    /**
     * Create multiple ports atomically.
     *
     * @param ports Port objects to create
     * @return Created Port objects
     */
    @throws(classOf[ConflictHttpException])
    @throws(classOf[NotFoundHttpException])
    def createPortBulk(@Nonnull ports: util.List[Port]): util.List[Port]

    /**
     * Delete a port. Nothing happens if the resource does not exist.
     *
     * @param id ID of the Port object to delete
     */
    @throws(classOf[ConflictHttpException])
    @throws(classOf[NotFoundHttpException])
    def deletePort(@Nonnull id: UUID)

    /**
     * Retrieve a port. Returns null if the resource does not exist.
     *
     * @param id ID of the Port object to delete
     * @return Port object
     */
    @throws(classOf[ConflictHttpException])
    @throws(classOf[NotFoundHttpException])
    def getPort(@Nonnull id: UUID): Port

    /**
     * Get all the ports.
     *
     * @return util.List of Port objects.
     */
    @throws(classOf[ConflictHttpException])
    @throws(classOf[NotFoundHttpException])
    def getPorts: util.List[Port]

    /**
     * Update a port.
     *
     * @param id ID of the Port object to update
     * @return Updated Port object
     */
    @throws(classOf[ConflictHttpException])
    @throws(classOf[NotFoundHttpException])
    def updatePort(@Nonnull id: UUID, @Nonnull port: Port): Port
}